modules/effects/arrow-effects-rx2-data/src/main/kotlin/arrow/effects/rx2/ObservableK.kt

```diff
@@ -79,29 +78,30 @@ data class ObservableK<A>(val observable: Observable<A>) : ObservableKOf<A>, Obs
    *  ```
    */
   fun <B> bracketCase(use: (A) -> ObservableKOf<B>, release: (A, ExitCase<Throwable>) -> ObservableKOf<Unit>): ObservableK<B> =
-    ObservableK(Observable.create<B> { emitter ->
-      observable.subscribe({ a ->
-        if (emitter.isDisposed) release(a, ExitCase.Canceled).fix().observable.subscribe({}, emitter::onError)
-        else try {
-          emitter.setDisposable(use(a).fix()
-            .flatMap { b -> release(a, ExitCase.Completed).fix().map { b } }
-            .handleErrorWith { e -> release(a, ExitCase.Error(e)).fix().flatMap { ObservableK.raiseError<B>(e) } }
-            .observable
-            .doOnDispose { release(a, ExitCase.Canceled).fix().observable.subscribe({}, emitter::onError) }
-            .subscribe(emitter::onNext, emitter::onError))
-        } catch (e: Throwable) {
-          if (NonFatal(e)) {
-            release(a, ExitCase.Error(e)).fix().observable.subscribe({
-              emitter.onError(e)
-            }, { e2 ->
-              emitter.onError(Platform.composeErrors(e, e2))
-            })
-          } else {
-            throw e
+    Observable.create<B> { emitter ->
+      val dispose =
+        handleErrorWith { t -> Observable.fromCallable { emitter.onError(t) }.flatMap { Observable.error<A>(t) }.k() }
+          .concatMap { a ->
+            if (emitter.isDisposed) {
+              release(a, ExitCase.Canceled).fix().observable.subscribe({}, emitter::onError)
+              Observable.never<B>().k()
+            } else {
+              defer { use(a) }
+                .value()
+                .doOnError { t: Throwable ->
+                  defer { release(a, ExitCase.Error(t.nonFatalOrThrow())) }.value().subscribe({ emitter.onError(t) }, emitter::onError)
```

Comment:
Where does this `OrThrow` go? Here it's swallowed by `defer`, which means `release` is not called but it's posted to `emitter::onError` immediately.

Extracting it outside of `defer` will it get swallowed by `create`, no?

I feel like we should drop `nonFatalOrThrow` here because I don't think `RxJava` takes into account fatal errors but I might be mistaken.

Reply:
It does IIRC.

Reply:
So is there still a need for us to interfere there or should we just pass the fatal error to Rx?

Reply:
Let's do it anyway.